dp/cloud/go/active_mode_controller/internal/message_generator/message_generator.go

```diff
@@ -87,6 +87,12 @@ func (m *messageGenerator) getPerCbsdMessageGenerator(cbsd *active_mode.Cbsd, no
 		cbsd.DbData.ShouldDeregister ||
 		cbsd.DesiredState == active_mode.CbsdState_Unregistered {
 		generator.sas = &sas.DeregistrationRequestGenerator{}
+	} else if cbsd.DbData.ShouldRelinquish {
+		if len(cbsd.Grants) == 0 {
+			generator.action = &relinquishMessageGenerator{}
+		} else {
+			generator.sas = &sas.RelinquishmentRequestGenerator{}
+		}
```

Comment:
this entire function is very hard to read because of all the if-ology. It'd be great to at least add a TODO item to make it more readable (e.g. separate the code out to sub-functions)

Reply:
Added todo